### DISCUSSION TITLE
Setting out_class in Hovernet pathology tutorial

### DISCUSSION THREAD

**Discussion by OmarAshkar** [Project-MONAI/tutorials#1180] [General]
**Describe the bug**
I am having a problem running the current hovernet tutorial using out_class other than 5. It gives me an incorrect tensor dimension error.  My data has only 4 classes

**To Reproduce**
1. Go to https://github.com/Project-MONAI/tutorials/tree/main/pathology/hovernet
2. Run training.py  with out_class = 4

**Comment by Nic-Ma** [comment]
Hi @KumoLiu ,

Could you please help share some comments about this question?

Thanks in advance.

**Comment by KumoLiu** [comment]
Hi @OmarAshkar, could you please share more information about the error so I could take a look at it? Thanks!

**Comment by OmarAshkar** [comment]
Hi @KumoLiu. Thanks for your help. I am pasting here the error message I am getting. 

```
  File "/opt/conda/lib/python3.8/site-packages/ignite/engine/engine.py", line 425, in _fire_event
    func(*first, *(event_args + others), **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/monai/handlers/checkpoint_saver.py", line 289, in exception_raised
    raise e
  File "/opt/conda/lib/python3.8/site-packages/ignite/engine/engine.py", line 1068, in _run_once_on_dataset_as_gen
    self.state.output = self._process_function(self, self.state.batch)
  File "/opt/conda/lib/python3.8/site-packages/monai/engines/trainer.py", line 228, in _iteration
    _compute_pred_loss()
  File "/opt/conda/lib/python3.8/site-packages/monai/engines/trainer.py", line 220, in _compute_pred_loss
    engine.state.output[Keys.LOSS] = engine.loss_function(engine.state.output[Keys.PRED], targets).mean()
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/monai/apps/pathology/losses/hovernet_loss.py", line 155, in forward
    self.dice(prediction[HoVerNetBranch.NC.value], target[HoVerNetBranch.NC.value]) * self.lambda_nc_dice
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/monai/losses/dice.py", line 157, in forward
    raise AssertionError(f"ground truth has different shape ({target.shape}) from input ({input.shape})")
AssertionError: ground truth has different shape (torch.Size([16, 5, 164, 164])) from input (torch.Size([16, 4, 164, 164]))
```

**Comment by KumoLiu** [comment]
Hi @OmarAshkar, seems the shape of your ground truth is incorrect. Did you consider the background class in your 4 classes? If not, I think `out_class` should set to 5. Thanks!

**Comment by OmarAshkar** [comment]
@KumoLiu. Yes, I am pretty sure my classes are total 4. Background included. But I will check also

**Comment by OmarAshkar** [comment]
Hi @KumoLiu. I have asserted the exported type_map has 4 classes using assert_array_equal. What else do you suggest to check the labels?

**Comment by KumoLiu** [comment]
Hi @OmarAshkar, from the error message, I could see that your label has 5 channels. So could you please check the shape of your type label? Something like this:
```
label_type = val_ds[case_num]["label_type"]
label_type_shape = label_type.shape
print(label_type_shape)
```
I think maybe you forgot to change `to_onehot=[5]` to `to_onehot=[4]` in the preprocessing.
https://github.com/Project-MONAI/tutorials/blob/c60dc6d9b12b0689d647a171a412a1c8bce5f7ff/pathology/hovernet/training.py#L217

Thanks!

**Reply by OmarAshkar** [reply]
@KumoLiu That did the trick thanks! May I ask another question? I have my very unbalanced classes and I wanted the dice metric to be weighted for each class separately then averaged. I have added reduction = "mean_batch", but still only single train_loss is printed while training

https://github.com/Project-MONAI/tutorials/blob/c60dc6d9b12b0689d647a171a412a1c8bce5f7ff/pathology/hovernet/training.py#L335-L336

How can I make the print for dice for each class separately before averaging?

Thanks alot!

**Reply by KumoLiu** [reply]
Hi @OmarAshkar, since the tutorial only calculates the binary dice from the segmentation branch, so even you set "mean_batch" you can only see a single metric is printed.
You can set `reduction = "mean_batch"` and `include_background=True` in the `key_train_metric`.
https://github.com/Project-MONAI/tutorials/blob/c60dc6d9b12b0689d647a171a412a1c8bce5f7ff/pathology/hovernet/training.py#L335-L338

Also added `to_onehot=2` in `AsDiscreted` in the post-processing.
https://github.com/Project-MONAI/tutorials/blob/c60dc6d9b12b0689d647a171a412a1c8bce5f7ff/pathology/hovernet/training.py#L271

Then you can see two dice metrics are printed, representing the foreground and background respectively. Something like:
`Epoch[1] Metrics -- train_dice: tensor([0.2784, 0.3410])`

Thanks!

**Reply by OmarAshkar** [reply]
Thanks @KumoLiu ! But I was wondering if there is a way to get a dice score for each one of the 4 classes?

**Reply by KumoLiu** [reply]
This way can show you the dice score for each class. Setting `reduction = "mean_batch"`. But if you want to see the dice score for the type map, you should calculate them since the tutorial didn't calculate it.

**Reply by OmarAshkar** [reply]
@KumoLiu Thanks alot! But I have one more clarification please. 

Isn't hovernet need to calculate the accuracy of the type map to train the classification branch? I was thinking there must be some sort of confusion matrix calculated to do so as described in the paper
